src/ansi-c/c_typecheck_code.cpp

```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
       << eom;
   }
 
-  // fatal errors
-  bool must_throw = false;
-
   if(condition.type().id() == ID_empty)
   {
-    must_throw = true;
-    error().source_location = condition.source_location();
-    error() << "void-typed expressions not allowed as assigns clause conditions"
-            << eom;
+    std::ostringstream error_message;
+    error_message
+      << "void-typed expressions not allowed as assigns clause conditions";
+    throw invalid_source_file_exceptiont{
+      error_message.str(), condition.source_location()};
   }
 
   if(has_subexpr(condition, [&](const exprt &subexpr) {
        return can_cast_expr<side_effect_exprt>(subexpr) &&
               !can_cast_expr<side_effect_expr_function_callt>(subexpr);
      }))
   {
-    must_throw = true;
-    error().source_location = condition.source_location();
-    error() << "side-effects not allowed in assigns clause conditions" << eom;
+    std::ostringstream error_message;
+    error_message << "side-effects not allowed in assigns clause conditions";
+    throw invalid_source_file_exceptiont{
+      error_message.str(), condition.source_location()};
   }
 
   if(has_subexpr(condition, ID_if))
   {
-    must_throw = true;
-    error().source_location = condition.source_location();
-    error() << "ternary expressions not allowed in assigns "
-               "clause conditions"
-            << eom;
+    std::ostringstream error_message;
+    error_message
+      << "ternary expressions not allowed in assigns clause conditions";
+    throw invalid_source_file_exceptiont{
+      error_message.str(), condition.source_location()};
   }
-
-  if(must_throw)
-    throw 0;
 }
 
-void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
+void c_typecheck_baset::typecheck_conditional_targets(
+  exprt::operandst &targets,
+  const std::function<void(exprt &)> typecheck_target,
+  const std::string &clause_type)
 {
-  // compute type
-  typecheck_expr(target);
+  exprt::operandst tmp;
 
-  // fatal errors
-  if(target.type().id() == ID_empty)
+  for(auto &target : targets)
   {
-    error().source_location = target.source_location();
-    error() << "void-typed expressions not allowed as assigns clause targets"
-            << eom;
-    throw 0;
-  }
+    if(!can_cast_expr<conditional_target_group_exprt>(target))
+    {
+      std::ostringstream error_message;
+      error_message << "expected a conditional target group expression in " +
+                         clause_type + "clause, found "
+                    << id2string(target.id());
```

Comment:
No need for `id2string` on this occasion.

Reply:
fixed